Library/Homebrew/os/linux/elf.rb

```diff
@@ -0,0 +1,151 @@
+module ELFShim
+  # See: https://en.wikipedia.org/wiki/Executable_and_Linkable_Format#File_header
+  MAGIC_OFFSET = 0
+  MAGIC_SIZE = 4
+  MAGIC_NUMBER = "\x7fELF".freeze
+
+  OSABI_OFFSET = 0x07
+  OSABI_SIZE = 1
+  OSABI_SYSTEM_V = 0
+  OSABI_LINUX = 3
+
+  TYPE_OFFSET = 0x10
+  TYPE_SIZE = 2
+  TYPE_EXECUTABLE = 2
+  TYPE_SHARED = 3
+
+  MACHINE_OFFSET = 0x12
+  MACHINE_SIZE = 2
+  MACHINE_I386 = 0x3
+  MACHINE_POWERPC = 0x14
+  MACHINE_ARM = 0x28
+  MACHINE_X86_64 = 0x62
+  MACHINE_AARCH64 = 0xB7
+
+  def elf?
+    return @elf if defined? @elf
+    return @elf = false unless read(MAGIC_SIZE, MAGIC_OFFSET) == MAGIC_NUMBER
+    # Check that this ELF file is for Linux or System V.
+    # OSABI is often set to 0 (System V), regardless of the target platform.
+    osabi = read(OSABI_SIZE, OSABI_OFFSET).unpack("C")[0]
+    @elf = osabi == OSABI_LINUX || osabi == OSABI_SYSTEM_V
+  end
+
+  def arch
+    return :dunno unless elf?
+    @arch ||= case read(MACHINE_SIZE, MACHINE_OFFSET).unpack("v")[0]
+    when MACHINE_I386 then :i386
+    when MACHINE_X86_64 then :x86_64
+    when MACHINE_POWERPC then :powerpc
+    when MACHINE_ARM then :arm
+    when MACHINE_AARCH64 then :arm64
+    else :dunno
+    end
+  end
+
+  def elf_type
+    return :dunno unless elf?
+    @elf_type ||= case read(TYPE_SIZE, TYPE_OFFSET).unpack("v")[0]
+    when TYPE_EXECUTABLE then :executable
+    when TYPE_SHARED then :dylib
+    else :dunno
+    end
+  end
+
+  def dylib?
+    elf_type == :dylib
+  end
+
+  def binary_executable?
+    elf_type == :executable
+  end
+
+  def dynamic_elf?
+    return @dynamic_elf if defined? @dynamic_elf
+    if which "readelf"
+      Utils.popen_read("readelf", "-l", to_path).include?(" DYNAMIC ")
+    elsif which "file"
+      !Utils.popen_read("file", "-L", "-b", to_path)[/dynamic|shared/].nil?
+    else
+      raise "Please install either readelf (from binutils) or file."
+    end
+  end
+
+  class Metadata
+    attr_reader :path, :dylib_id, :dylibs
+
+    def initialize(path)
+      @path = path
+      @dylibs = []
+      @dylib_id, needed = needed_libraries path
+      return if needed.empty?
```

Comment:
Stick a newline after each `return`; makes this easier to follow/scan (and consider doing this elsewhere in the file).